File: apps/ios-wiki-demo/App/Model/CorpusSidecar.swift
import Foundation

/// `corpus.json` inside the 008 index directory (008 data-model "Corpus identity") — the
/// fields the About screen shows; unknown keys are ignored.
struct CorpusSidecar: Codable, Equatable {
    struct Snapshot: Codable, Equatable {
        let edition: String
        let snapshotDate: String
        enum CodingKeys: String, CodingKey { case edition; case snapshotDate = "snapshot_date" }
    }
    struct Counts: Codable, Equatable {
        let articles: Int
        let selected: Int
        let passages: Int
    }
    let schemaVersion: Int
    let corpusIdentity: String
    let snapshot: Snapshot
    /// The article limit of a slice (`xtriever wiki build --limit N`); absent on the whole edition.
    let partial: Int?
    let counts: Counts
    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case corpusIdentity = "corpus_identity"
        case snapshot, partial, counts
    }

    /// About's line for a slice; nil on the whole edition.
    var sliceNote: String? { partial.map { "first \($0) articles, not the whole edition" } }

    /// A measured run's `corpus`: a slice's numbers must never read as the whole edition's.
    var runLabel: String { partial.map { "wikipedia, first \($0) articles" } ?? "wikipedia" }

    static func load(from indexDir: URL) throws -> CorpusSidecar {
        try JSONDecoder().decode(CorpusSidecar.self, from: Data(contentsOf: indexDir.appendingPathComponent("corpus.json")))
    }
}

File: apps/android-wiki-demo/src/main/kotlin/dev/xtriever/demo/CorpusSidecar.kt
package dev.xtriever.demo

import org.json.JSONObject
import java.io.File

/**
 * The Feature 008 corpus sidecar, read as it stands (Feature 025, spec FR-009). The application
 * computes nothing from it: every line is what the build recorded.
 */
data class CorpusSidecar(val facts: Map<String, String>) {
    companion object {
        fun read(file: File): CorpusSidecar {
            if (!file.isFile) return CorpusSidecar(emptyMap())
            return parse(JSONObject(file.readText()))
        }

        fun parse(json: JSONObject): CorpusSidecar {
            val counts = json.optJSONObject("counts")
            val snapshot = json.optJSONObject("snapshot")
            return CorpusSidecar(
                buildMap {
                    snapshot?.let {
                        put("corpus", "Simple English Wikipedia (${it.optString("edition")})")
                        put("snapshot", it.optString("snapshot_date"))
                    }
                    counts?.let {
                        put("articles", "${it.optInt("articles")} read, ${it.optInt("selected")} selected")
                        put("passages", it.optInt("passages").toString())
                    }
                    if (json.has("partial") && !json.isNull("partial")) {
                        put("partial", "first ${json.opt("partial")} articles")
                    }
                    put("corpus identity", json.optString("corpus_identity"))
                },
            )
        }
    }
}
